### PR TITLE
Support mixed MX element dtype in `mx_mm` function and `MXLinear`.

### PR DESCRIPTION
**Pull request by balancap**
Following the MXFP and quantization literature, it is useful to support different element dtypes for activations, weights and gradients. This PR is simply adding a more general interface to `mx_mm`. A similar choice could be done with `MXLinear`

General issue: #1666

### CODE REVIEW

File: torchao/prototype/mx_formats/mx_linear.py [hunk 1]
```diff
@@ -23,25 +23,31 @@ class mx_mm(torch.autograd.Function):
     # 1.       input @ weight_t    = output     (forward pass)
     # 2. grad_output @ weight      = grad_input (backward pass)
     # 3.     input_t @ grad_output = grad_weight (backward pass)
+    # 
+    # input, weight and grad_output have each their own MX element dtype.
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_linear.py:27]: nit: "can have"?

**Comment by vkuzo** [comment]
this makes sense, it would be great to cover with a test

the easiest place to test it would be here (https://github.com/pytorch/ao/blob/8afd10ed4b22b3cabd80184062c4ad58001bc68a/test/prototype/mx_formats/test_mx_linear.py#L47), and that requires adding this to `MXLinear`.  Would you be interested in doing that in this PR?

by the way, https://github.com/pytorch/pytorch/issues/146414 outlines bringing MX dtypes to PyTorch core, and we plan to evolve `torchao/prototype/mx_formats/` accordingly

File: torchao/prototype/mx_formats/mx_linear.py (hunk 1 above)
Comment (balancap) [reply torchao/prototype/mx_formats/mx_linear.py]: Done

**Comment by balancap** [comment]
I added the support of this feature in `MXLinear` too. In order to avoid breaking the interface (and keeping things simple in the single dtype case), you can now pass either a single element dtype or a tuple of 3.

I expanded the coverage in the test you mentioned (plus a small test on the factory side to check the 2 cases above are working properly).

Thanks for the link on PyTorch MX plan :+1: I would assume that the MX "simulated" mode is going to stay in TorchAO for some time, as it is very useful for testing + getting ready for MX hardware until it is widely available.

File: torchao/prototype/mx_formats/mx_linear.py [hunk 2]
```diff
@@ -87,13 +97,20 @@ class MXLinear(torch.nn.Linear):
...
+    Input, weight and grad_output can have each their own MX element dtype
+    by passing a tuple of `elem_dtype` to the factory method `from_float`.
     """
 
     @classmethod
     @torch.no_grad()
     def from_float(cls, mod, elem_dtype, block_size):
         mod.__class__ = MXLinear
-        mod.elem_dtype = elem_dtype
+        # Single element dtype passed for input, weight and gradient.
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_linear.py:109]: nit: can we do

```
def from_float(
    ...,
    elem_dtype=torch.float8_e4m3fn,
    elem_dtype_weight_override=None,
    elem_dtype_grad_output_override=None,
    ...
): ...
```

we plan to create a proper config object for this in the future, but for now would be good to keep things simple and avoid mixing types in the API (such as dtype vs tuple)

**Comment by vkuzo** [comment]
> I would assume that the MX "simulated" mode is going to stay in TorchAO for some time, as it is very useful for testing + getting ready for MX hardware until it is widely available.

yep! great to hear this is useful.

File: torchao/prototype/mx_formats/mx_linear.py (hunk 2 above)
Comment (balancap) [reply torchao/prototype/mx_formats/mx_linear.py:109]: Should I then enforce named argument in `MXLinear.from_float` and `swap_linear_with_mx_linear` for `block_size` and `filter_fn`? And have a default `block_size=32`?

Comment (vkuzo) [reply torchao/prototype/mx_formats/mx_linear.py:109]: sounds reasonable!

Comment (balancap) [reply torchao/prototype/mx_formats/mx_linear.py]: Just pushed a fix commit with:
```python
def from_float(
        cls,
        mod,
        elem_dtype,
        elem_dtype_weight_override=None,
        elem_dtype_grad_output_override=None,
        *,
        block_size=32,
    ):
```
and similarly for `swap_linear_with_mx_linear`

**Review by vkuzo — APPROVED** [review]
looks good, thank you!  Please feel free to merge if CI is green.

note: we will likely change the UX of this workflow in the near future (add a top-level config, etc) as we add Blackwell support, we'll make sure to keep these options in the new UX!

**Comment by balancap** [comment]
@vkuzo Unfortunately, the H100 Float8 test runner seems to have had an issue starting.

**Comment by vkuzo** [comment]
failure is transient, I merged it.  Thank you!
